### ISSUE TITLE
Training bubble between epochs when finetuning GR00T N1.5 with LeRobot dataset

### ISSUE THREAD

**Issue by KingsleyLiu-NV** [NVIDIA/Isaac-GR00T#345]
**Description**

When finetuning GR00T N1.5 using a LeRobot-compatible dataset, I noticed a bubble (idle time gap) between epochs. This overhead may be related to data loading, dataset re-initialization, or other input pipeline inefficiencies.

<img width="2112" height="1220" alt="Image" src="https://github.com/user-attachments/assets/52a1544d-5e44-49e7-9636-54d089da32cb" />

**Expected Behavior**

Training should seamlessly continue between epochs without significant idle gaps. Ideally, data loading or shuffling should overlap with training so that GPU utilization remains consistently high across epochs.

**Steps to reproduce**

```bash
nsys profile -o nsys-profiles/finetune-n1.5-3b-bs32 \
python ./scripts/gr00t_finetune.py \
  --dataset-path ./demo_data/robot_sim.PickNPlace \
  --report-to tensorboard \
  --max-steps 100 \
  --output-dir ckpts \
  --base-model-path /workspace/GR00T-N1.5-3B \
  --num-gpus 1 \
  --batch-size 32
```

**Version and Platform**

Python 3.12.3

Isaac-GR00T main branch

NVIDIA GPU

@xwjiang2010 @DrJimFan Thanks for the help!

**Comment by youliangtan** [comment]
Thanks @KingsleyLiu-NV for this analysis. This is an known issue. Unfortunately, our current native dataloader is video decoding IO bounded. We have an internal sharded implementation which is more complicated, can use up alot of cpu mem. It comes with its own limitations due to IID-ness of dataset etc.

We are happy to take in feedback for better advice on this too. =)

**Comment by xupinjie** [comment]
Hello @youliangtan，thanks for your advise. Based on @KingsleyLiu-NV 's configuration, I conducted some tests on the dataloader using an H100 and a 32-core Intel(R) Xeon(R) Silver 4314 CPU with the LeRobot-compatible dataset. The results show that the `get_batch_samples` function accounts for about 2% of the main process time (10ms out of 370ms). When the dataloader uses 8 workers, data load overhead is effectively hidden by the training workload. No I/O bottleneck was observed. Could you please advise on what factors might be causing this? Also, could it be related to the dataset used (LeRobot-compatible dataset), since the videos are relatively small (256x256 resolution, 1.1 MB)?

<img width="884" height="639" alt="Image" src="https://github.com/user-attachments/assets/69ead4d0-c904-417e-bf0b-094c36d4d333" />

**Comment by xwjiang2010** [comment]
I think it depends on the size/resolution of the videos, the codec used, the level of video processing etc.
But in practice, I do see dataloading being a major bottleneck of the training.

@xupinjie what would you say result in the difference in your observation and @KingsleyLiu-NV 's? I am curious to learn.

**Comment by xupinjie** [comment]
@xwjiang2010 @youliangtan , I believe the bubble in @KingsleyLiu-NV’s timeline occurs between epochs. If dataloading is a major bottleneck in training, the bubble should appear between iterations. To investigate, I ran some experiments attempting to identify the bubble between iterations.

Thanks for @xwjiang2010  advice; it might be caused by video resolution. Could you provide some data from real training sessions or share details about the actual videos you’re using (such as resolution, codec, or other relevant info)? I really appreciate your help.

**Comment by xwjiang2010** [comment]
Hmm, are we sure the bubble happens at epoch boundary?
How far into training do we see this bubble? Imagine that we have 50 trajectories and each trajectory is like 20s. Frequency is 50ms. So we have 400 frames per trajectory. All together we have 50 * 1K = 50K timesteps. Assume a batch size of 64 * 8GPUs = 512. You need 50K / 512 iterations to hit epoch boundary, i.e. 100 iterations. Is this what we are seeing when the slow happens?

My personal experience is usually I train a few steps with relatively ok speed, and then stall for quite some time without any training progress. I hypothesis this is because I have 12 dataloader worker/GPU and that can give me say 12 batches. Then I have to wait some time for another 12 batches to be available for my GPU. So there is burst and stall interleaving pattern. 

You can try my training by following this: https://github.com/NVIDIA/Isaac-GR00T/pull/357
Maybe try LIBERO-spatial.

**Comment by KingsleyLiu-NV** [comment]
@xwjiang2010 I am sure the epoch-boundary bubble is NOT related to video decoding bottleneck. After diving into the code, I find the the epoch iterator is reconstructed at the beginning of each epoch: [trainer.py#L2606](https://github.com/huggingface/transformers/blob/main/src/transformers/trainer.py#L2606) and sampler is on a epoch basis: [trainer.py#L46-L52](https://github.com/NVIDIA/Isaac-GR00T/blob/main/gr00t/experiment/trainer.py#L46-L52). Due to this, the initial overhead of reading parquet files cannot be hidden when a new epoch starts.

I did some experiments after modifying the code to have persistent epoch iterator, and the epoch-boundary bubble disappeared using the same training config and dataset.

```bash
class BaseSampler(Sampler):
    def __iter__(self):
        # """🚀 INFINITE SAMPLER: Never stops, cycles through epochs forever!"""
        current_epoch = self.epoch
        
        while True: 
            print(f"🚀 BaseSampler: Starting epoch {current_epoch}")
            
            if self.shuffle:
                g = torch.Generator()
                g.manual_seed(self.seed + current_epoch)
                indices = torch.randperm(len(self.data_source), generator=g).tolist()
            else:
                indices = list(range(len(self.data_source)))
            
            for idx in indices:
                yield idx
            
            current_epoch += 1
            print(f"🔄 BaseSampler: Seamlessly switching to epoch {current_epoch}")
```

```bash
            # 🚀 PERSISTENT ITERATOR: Only create once, reuse forever!
            if not hasattr(self, '_persistent_epoch_iterator'):
                print(f"🚀 Creating PERSISTENT epoch_iterator (one-time cost)")
                self._persistent_epoch_iterator = iter(epoch_dataloader)
                print(f"🕐 Initial iter(epoch_dataloader) took {time.time()-t1:.3f}s")
            else:
                print(f"🚀 REUSING persistent epoch_iterator (zero cost!)")
                print(f"🕐 iter(epoch_dataloader) took 0.000s (cached)")
            
            epoch_iterator = self._persistent_epoch_iterator
```

**Comment by wenjing-nv** [comment]
Hi @xwjiang2010  @xupinjie are profiling the GPU video decoding because @youliangtan commented [here](https://github.com/NVIDIA/Isaac-GR00T/issues/345#issuecomment-3264778142) that  video decoding IO bounded is an issue for you. So we chime in to solve that. Thanks
